src/cert/x509.rs

```diff
@@ -0,0 +1,278 @@
+            Ok((algo, aparams))
+        })?;
+
+        der::bits_total(buf)?.read_all(Error::BadEncoding, |buf| match algo {
+            oid::RSA_ENCRYPTION => {
+                aparams.read_all(Error::BadEncoding, der::null)?;
+                der::tagged(Tag::SEQUENCE, buf, |buf| {
+                    let mut modulus = der::uint(buf)?.as_slice_less_safe();
+                    // DER inserts a leading zero sometimes (to disambiguate
+                    // negative integers) so we need to remove it.
+                    if modulus[0] == 0 {
+                        modulus = &modulus[1..];
+                    }
+                    let mut exponent = der::uint(buf)?.as_slice_less_safe();
+                    if exponent[0] == 0 {
+                        exponent = &exponent[1..];
+                    }
+                    Ok(PublicKeyParams::Rsa { modulus, exponent })
+                })
+            }
+            _ => Err(Error::UnknownAlgorithm),
+        })
+    })?;
+
+    // We don't care about the UIDs at all.
+    let _issuer_uid = der::opt(Tag::context_specific(1), buf)?;
+    let _subject_uid = der::opt(Tag::context_specific(2), buf)?;
+
+    // Extensions are mandatory.
+    let mut extns = Extensions::default();
+    der::tagged(Tag::context_specific(3), buf, |buf| {
+        der::tagged(Tag::SEQUENCE, buf, |buf| {
+            while !buf.at_end() {
+                parse_extn(buf, &mut extns)?;
+            }
+            Ok(())
+        })
+    })?;
+
+    let is_cert_sign = match extns.is_cert_sign {
+        Some(b) => b,
+        _ => return Err(Error::BadEncoding),
+    };
+
+    let is_ca = match &extns.basic_constraints {
+        Some(bc) => bc.is_ca,
+        _ => false,
+    };
+
+    // CA certificates must always specify keyCertSign as a valid usage. To
+    // fail to do so is a syntax error.
+    if is_ca != is_cert_sign {
+        return Err(Error::BadEncoding);
+    }
+
+    Ok(Cert {
+        format,
+        issuer,
+        subject,
+        subject_key,
+        basic_constraints: extns.basic_constraints,
+        is_cert_sign,
+    })
+}
+
+#[derive(Default)]
+struct Extensions {
+    basic_constraints: Option<cert::BasicConstraints>,
+    is_cert_sign: Option<bool>,
+}
+
+fn parse_extn(
+    buf: &mut untrusted::Reader,
+    extns: &mut Extensions,
+) -> Result<(), Error> {
+    der::tagged(Tag::SEQUENCE, buf, |buf| {
+        let oid = der::oid(buf)?;
+        let is_critical = der::opt_bool(buf)?.unwrap_or(false);
+        der::tagged(Tag::OCTET_STRING, buf, |buf| match oid {
+            oid::KEY_USAGE => {
+                if extns.is_cert_sign.is_some() {
+                    return Err(Error::BadEncoding);
+                }
+
+                // Note that the bits in a `BIT STRING` are big endian
+                // within bytes, so we need to "shift from the left"
+                // instead.
+                const CERT_SIGN_MASK: u16 = 1 << (15 - 5);
```

Comment:
key usage bits as defined in RFC 5280 section 4.2.1.3? (add a comment?)

Reply:
Also, would it make sense from a namespace perspective to declare this const in the `Extensions` struct?

Reply:
Done. I don't think moving it in is super necessary unless we start propagating it somehow, and even then I think it's unlikely.